fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,576 @@
+        }
+
+        return createPropertySetByDistribution(new HashDistributionSpec(
+                new HashDistributionDesc(olapDistributionSpec.getShuffleColumns(),
+                        HashDistributionDesc.SourceType.SHUFFLE_LOCAL), physicalPropertyInfo));
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalTopN(PhysicalTopNOperator topN, ExpressionContext context) {
+        PhysicalPropertySet outputProperty;
+        if (topN.getSortPhase().isFinal()) {
+            if (topN.isSplit()) {
+                DistributionSpec distributionSpec = DistributionSpec.createGatherDistributionSpec();
+                DistributionProperty distributionProperty = new DistributionProperty(distributionSpec);
+                SortProperty sortProperty = new SortProperty(topN.getOrderSpec());
+                outputProperty = new PhysicalPropertySet(distributionProperty, sortProperty);
+            } else {
+                outputProperty = new PhysicalPropertySet(new SortProperty(topN.getOrderSpec()));
+            }
+        } else {
+            outputProperty = PhysicalPropertySet.EMPTY;
+        }
+        return outputProperty;
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalAnalytic(PhysicalWindowOperator node, ExpressionContext context) {
+        Preconditions.checkState(childrenOutputProperties.size() == 1);
+        List<Integer> partitionColumnRefSet = new ArrayList<>();
+
+        node.getPartitionExpressions().forEach(e -> {
+            partitionColumnRefSet
+                    .addAll(Arrays.stream(e.getUsedColumns().getColumnIds()).boxed().collect(Collectors.toList()));
+        });
+
+        SortProperty sortProperty = new SortProperty(new OrderSpec(node.getEnforceOrderBy()));
+
+        DistributionProperty distributionProperty;
+        if (partitionColumnRefSet.isEmpty()) {
+            distributionProperty = new DistributionProperty(DistributionSpec.createGatherDistributionSpec());
+        } else {
+            // Use child distribution
+            distributionProperty = childrenOutputProperties.get(0).getDistributionProperty();
+        }
+        return new PhysicalPropertySet(distributionProperty, sortProperty);
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalFilter(PhysicalFilterOperator node, ExpressionContext context) {
+        Preconditions.checkState(this.childrenOutputProperties.size() == 1);
+        return childrenOutputProperties.get(0);
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalTableFunction(PhysicalTableFunctionOperator node,
+                                                          ExpressionContext context) {
+        Preconditions.checkState(this.childrenOutputProperties.size() == 1);
+        return childrenOutputProperties.get(0);
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalLimit(PhysicalLimitOperator node, ExpressionContext context) {
+        return createLimitGatherProperty(node.getLimit());
```

Comment:
Why limit operator must output gather property ?

Reply:
Because it requires gather property